fix graal native image - improvement

zeitlinger opened the pull request:
We broke native image and didn't realize it because 

- main was broken at the time due to spring 7 latest deps
- it took more that 1 PR to fix latest deps and thus didn't see the failure

laurit commented:
I created https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/15311 that uses a slightly different approach just to test out whether it would work. Noticed that for webflux instrumentation the header getter logic is duplicated for client and server instrumentation, fixed that too. The approach taken in this PR is fine too and I might even prefer it. Though first all tests would need to pass.

zeitlinger commented:
> I created #15311 that uses a slightly different approach just to test out whether it would work. Noticed that for webflux instrumentation the header getter logic is duplicated for client and server instrumentation, fixed that too. The approach taken in this PR is fine too and I might even prefer it. Though first all tests would need to pass.

great - I'm also trying to fix this PR so that we can choose.

- I've also opted not to ignore exceptions (which has already made it hard to fix the PR) 
- added explicit check for spring 7 instead
- not removed duplication yet

zeitlinger commented:
@laurit should I rebase - or just close this PR?

laurit commented on instrumentation/spring/spring-web/spring-web-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/web/v3_1/HeaderUtil.java:
```diff
@@ -34,20 +48,17 @@ private static MethodHandle findGetHeadersMethod(MethodType methodType) {
   // errorprone BadInstanceof warning since errorpone is not aware that this instanceof check does
   // not pass with spring web 7.0+
-  @SuppressWarnings({"unchecked", "BadInstanceof"})
+  @SuppressWarnings("unchecked") // casting MethodHandle.invoke result
```
the comment above is not valid any more

laurit commented on instrumentation/spring/spring-web/spring-web-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/web/v3_1/HeaderUtil.java:
```diff
@@ -6,26 +6,40 @@
+      @Nullable MethodHandle first, Supplier<? extends MethodHandle> supplier) {
+    return (first != null)
+        ? first
+        : requireNonNull(supplier.get(), "Could not find suitable get method on HttpHeaders");
```
You have to be really careful with throwing exceptions like this. For library instrumentations it isn't too bad, but in javaagent code you shouldn't throw exceptions like this (keep in mind that javaagent instrumentations sometimes use library instrumentations). In javaagent instrumentations we try not to break applications. If there are changes in the instrumented library that are not compatible with the instrumentation we disable the instrumentation instead of letting it fail. This is handled by muzzle. In javaagent instrumentations you should not fail if the methods that you need don't exist.

zeitlinger commented:
@laurit thanks for the feedback

I also added nullaway to make the new code more maintainable
